### PR TITLE
Introduce a `ConnectionTarget` enum

### PR DESCRIPTION
**Pull request by karwa**
I added this as part of the WebURL port, because it can provide specific host kinds and doesn't need to re-parse the hostname, but I think it is generally a useful thing to model.

Currently, `ConnectionPool.Key` stores a `host`, `port`, and `socketPath` string, even though only one of (host + port) and (socketPath) can be populated. This lends itself well to an enum. Additionally, we can extract what kind of host we're dealing with, so we don't need to re-parse the hostname later.

One thing I'd like comments on is whether this should perhaps be moved up to `Request`. This would also allows us to stop pretending that unix domain sockets have ports. As I see it, this should really be `Request`'s business; a part of how it extracts information from the URL should be parsing the hostname string (perhaps using APIs from the URL type, if it provides that). 

The async redesign simplifies `Request` somewhat, but I couldn't see any mention of hosts or hostnames on the new request type. One idea worth exploring would be making this public API as part of that redesign.

### CODE REVIEW

**Comment by karwa** [comment]
OK, I just did it and moved it up in to `Request`. I couldn't think of a good reason not to.

**Comment by karwa** [comment]
One idea I'd like to float is potentially naming this `ConnectionTarget`, since I'm not sure unix sockets really count as "hosts" in the typical sense (`Request.host` currently returns an empty string for them, even though HTTP requests to empty hosts are technically invalid). 

Also, it would avoid overloading the term "host" too much. It already clashes with `Request.host`, meaning it requiring an underscored or other suboptimal name.

**Comment by Lukasa** [comment]
`ConnectionTarget` is a good name, I'm in favour of that amendment.

**Comment by karwa** [comment]
OK, I've changed it to `ConnectionTarget`, and added tests for the quirky embedded IPv4 addresses.

**Comment by dnadoba** [comment]
Did a quick read through the changes and it is a nice cleanup. Thanks! Will do a proper review on monday.
I'm also refactoring some related stuff like `scheme` and moving the destructured URL into a separate type. If you are interested you can have a look at this branch https://github.com/dnadoba/async-http-client/tree/dn-scheme 
Still work in progress and naming is not final.
Nevertheless, I think we should land your changes first.

File: Sources/AsyncHTTPClient/ConnectionPool.swift
Comment (dnadoba) [inline Sources/AsyncHTTPClient/ConnectionPool.swift:23]: FYI: For now its probably okay to use `ConnectionTarget` here and it also does not change the behaviour compared to the old implementation. However, I think in the future we should probably use a different type that normalises the `ConnectionTarget`. E.g. the host should be compare case insensitive and different spellings of the same IP address should also result in the same `ConnectionPool.Key`.  Otherwise we end up with two different connections if a user wants to connect to `https://apple.com` and `https://APPLE.com` which we do not want.

**Review by dnadoba — APPROVED** [review]
Great patch! Looks good to me modulo one nit. We should still wait for @fabianfett review until we merge it.

w.r.t.:
> The async redesign simplifies Request somewhat, but I couldn't see any mention of hosts or hostnames on the new request type. One idea worth exploring would be making this public API as part of that redesign.

The new Request type, called `HTTPClientRequest`, does not do any parsing or validation. It also just accepts a url as a `String`. Parsing, destructering and validation of the url is all delayed until execution. The advantage is that the initialiser of `HTTPClientRequest` is non-throwing and all properties are `var`s. Therefore we will not even have `ConnectionTarget` as an internal property on `HTTPClientRequest`.

My plan is to introduce a new internal type called `PreparedRequest` which is created as part of executing the request. This type will be the new home for all properties which are computed as part of the current initialiser of `HTTPClient.Request` and the validation step which is currently also executed before sending out a `HTTPClient.Request`.

File: Sources/AsyncHTTPClient/ConnectionPool/ChannelHandler/HTTP1ProxyConnectHandler.swift
Comment (Lukasa) [inline Sources/AsyncHTTPClient/ConnectionPool/ChannelHandler/HTTP1ProxyConnectHandler.swift:49]: Minor nit: I'd like to put `convenience init` on the same line.

**Review by Lukasa — APPROVED** [review]
One nit, and I'd like to see @dnadoba's nits addressed, but otherwise this is great. Nice patch @karwa!

**Review by fabianfett — APPROVED** [review]

**Comment by dnadoba** [comment]
@karwa I merge it now and fix the nits in a follow up PR quickly because I want to use them in another PR. Thanks again for this patch!

**Comment by karwa** [comment]
Thanks @dnadoba! I was planning to fix in the evening yesterday, but was too tired 😴

**Pull request merged**
